lib/sensors/MoistureSensor.cpp

```diff
@@ -0,0 +1,93 @@
+                                                                            //it's just what I found through my own testing. It sometimes fluctuates on its own, so it's hard to trust :(
+
+constexpr int Sensor_Moisture_Function = 0x10;                              //Function address registers for moisture and temperature modules        
+constexpr int Sensor_Temp_Function = 0x04;  
+
+MoistureSensor::MoistureSensor(PinName sda, PinName scl) : i2c_(sda, scl){}
+
+bool MoistureSensor::Is_Initialized(){
+    if(this->Read_HW_ID() == Sensor_HW_ID_Code){                            //compare received HW ID Code to correct one
+        return true;
+    }
+
+    return false;
+}
+
+int MoistureSensor::Read_HW_ID(){
+    char cmd[2];
+    cmd[0] = Sensor_Status_Base;
+    cmd[1] = Sensor_Status_HW_ID;
+
+    char check[1];
+
+    i2c_.write(Sensor_I2C_Address, cmd, 2);                                 //initialize registers for checking device ID
+
+    i2c_.read(Sensor_I2C_Address, check, 1);                                //read device ID
+
+    return check[0];
+}
+
+uint16_t MoistureSensor::Read_Moisture(){
+    /*
+    if(!(this->Is_Initialized())){                                          //checks if device is initialized, returns 65534 if there is an issue
+        return 65534;                                                       //this is commented out because the HW_ID reading isn't consistent currently
+    }
+    */
+
+    char cmd[2];
+    cmd[0] = Sensor_Moisture_Base;
+    cmd[1] = Sensor_Moisture_Function;
+
+    char buf[2];
+    uint16_t ret = 65535;
+
+    do{
+        wait_ms(1);
+
+        i2c_.write(Sensor_I2C_Address, cmd, 2);                             //initialize registers for reading moisture
+
+        wait_ms(1);
+
+        i2c_.read(Sensor_I2C_Address, buf, 2);                              //read moisture
+
+        ret = ((uint16_t)buf[0] << 8 | buf[1]);                             //concatenate bytes together
+
+    } while(ret == 65535);                                                  //repeat until value has been measured
+
+    return ret;
+}
+
+float MoistureSensor::Read_Temperature(){
+    /*
+    if(!(this->Is_Initialized())){                                          //checks if device is initialized, returns -273.0 if there is an issue
+        return -273.0;                                                      //this is commented out because the HW_ID reading isn't consistent currently
+    }
+    */
+
+    char cmd[2];
+    cmd[0] = Sensor_Status_Base;
+    cmd[1] = Sensor_Temp_Function;
+
+    char buf[4];
+
+    i2c_.write(Sensor_I2C_Address, cmd, 2);                                 //initialize registers for reading temperature
+
+    wait_ms(1);
+
+    i2c_.read(Sensor_I2C_Address, buf, 4);                                  //read temp
+
+    int32_t ret = ((uint32_t)buf[0] << 24) | ((uint32_t)buf[1] << 16) |     //concatenate bytes together
```

Comment:
this i fine but a cleaner way to do this is with unions

Reply:
unless its reverse endian

Reply:
tbh, this is completely ripped from the given Arduino code (it looked good enough to me lol)